src/cascadia/TerminalSettingsModel/ApplicationState.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "ApplicationState.h"
+#include "CascadiaSettings.h"
+#include "ApplicationState.g.cpp"
+
+#include "JsonUtils.h"
+#include "FileUtils.h"
+
+static constexpr std::string_view CloseAllTabsWarningDismissedKey{ "closeAllTabsWarningDismissed" };
+static constexpr std::string_view LargePasteWarningDismissedKey{ "largePasteWarningDismissed" };
+static constexpr std::string_view MultiLinePasteWarningDismissedKey{ "multiLinePasteWarningDismissed" };
+
+using namespace ::Microsoft::Terminal::Settings::Model;
+
+// Function Description:
+// - Returns the path to the state file that lives next to the user's settings.
+static std::filesystem::path _statePath()
+{
+    const auto statePath{ GetBaseSettingsPath() / L"state.json" };
+    return statePath;
+}
+
+namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+{
+    // Function Description:
+    // - Returns a mutex and storage location for the application-global ApplicationState object
+    static std::tuple<std::reference_wrapper<std::mutex>, std::reference_wrapper<winrt::com_ptr<ApplicationState>>> _getStaticStorage()
+    {
+        static std::mutex mutex{};
+        static winrt::com_ptr<ApplicationState> storage{ nullptr };
+        return std::make_tuple(std::ref(mutex), std::ref(storage));
+    }
+
+    // Method Description:
+    // - Returns the application-global ApplicationState object.
+    Microsoft::Terminal::Settings::Model::ApplicationState ApplicationState::GetForCurrentApp()
+    {
+        auto [mtx, state] = _getStaticStorage();
+        std::lock_guard<std::mutex> lock{ mtx };
+        auto& s{ state.get() };
+        if (s && !s->_invalidated)
+        {
+            return *s;
+        }
+
+        auto newState{ winrt::make_self<ApplicationState>(_statePath()) };
+        newState->_load();
+        s = newState;
+        return *newState;
+    }
+
+    // Method Description:
+    // - Deserializes a state.json document into an ApplicationState.
+    // - *ANY* errors will result in the creation of a new empty state
+    void ApplicationState::_load()
+    {
+        wil::unique_hfile hFile{ CreateFileW(_path.c_str(),
+                                             GENERIC_READ,
+                                             FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                             nullptr,
+                                             OPEN_EXISTING,
+                                             FILE_ATTRIBUTE_NORMAL,
+                                             nullptr) };

```

Comment:
Double checking (because I'm not very familiar with this API), if this file doesn't exist, it'll create it, right?